File: StreamTriad/C/streamTriad.c
/*
 * =======================================================================================
 *
 *      Author:   Jan Eitzinger (je), jan.eitzinger@fau.de
 *      Copyright (c) 2019 RRZE, University Erlangen-Nuremberg
 *
 *      Permission is hereby granted, free of charge, to any person obtaining a copy
 *      of this software and associated documentation files (the "Software"), to deal
 *      in the Software without restriction, including without limitation the rights
 *      to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *      copies of the Software, and to permit persons to whom the Software is
 *      furnished to do so, subject to the following conditions:
 *
 *      The above copyright notice and this permission notice shall be included in all
 *      copies or substantial portions of the Software.
 *
 *      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *      IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *      FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *      AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *      LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *      OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *      SOFTWARE.
 *
 * =======================================================================================
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include <limits.h>
#include <float.h>

#ifdef _OPENMP
#include <omp.h>
#endif

#define SIZE 20000000ull
#define NTIMES 3
#define ARRAY_ALIGNMENT 64
#define HLINE "----------------------------------------------------------------------------\n"

#ifndef MIN
#define MIN(x,y) ((x)<(y)?(x):(y))
#endif
#ifndef MAX
#define MAX(x,y) ((x)>(y)?(x):(y))
#endif
#ifndef ABS
#define ABS(a) ((a) >= 0 ? (a) : -(a))
#endif

extern double striad_seq(double*, const double*, const double*, const double*, int, int);
extern double striad_tp(double*, const double*, const double*, const double*, int, int);
extern double striad_ws(double*, const double*, const double*, const double*, int, int);
extern double getTimeStamp();

typedef double (*testFunc)(double*, const double*, const double*, const double*, int, int);

int main (int argc, char** argv)
{
    size_t bytesPerWord = sizeof(double);
    size_t N;
    int type;
    size_t iter = 1;
    size_t scale = 1;
    double *a, *b, *c, *d;
    double E, S;
    double	avgtime, maxtime, mintime;
    double times[NTIMES];
    double dataSize;
    testFunc func;
    char* testname;

    if ( argc > 2 ) {
        type = atoi(argv[1]);
        N = atoi(argv[2]);
    } else {
        printf("Usage: %s <test type>  <N>\n",argv[0]);
        printf("Test types: 0 - sequential, 1 - OpenMP throughput, 2 - OpenMP worksharing\n");
        exit(EXIT_SUCCESS);
    }

    switch ( type ) {
        case 0:
            func = striad_seq;
            testname = "striad_seq";
            break;
        case 1:
            func = striad_tp;
            testname = "striad_tp";
#ifdef _OPENMP
#pragma omp parallel
            {
#pragma omp single
                scale = omp_get_num_threads();
            }
#endif
            break;
        case 2:
            func = striad_ws;
            testname = "striad_ws";
            break;
        default:
            printf("Unknown test type: %d\n", type);
            exit(EXIT_FAILURE);
    }

    posix_memalign((void**) &a, ARRAY_ALIGNMENT, N * bytesPerWord );
    posix_memalign((void**) &b, ARRAY_ALIGNMENT, N * bytesPerWord );
    posix_memalign((void**) &c, ARRAY_ALIGNMENT, N * bytesPerWord );
    posix_memalign((void**) &d, ARRAY_ALIGNMENT, N * bytesPerWord );

        avgtime = 0;
        maxtime = 0;
        mintime = FLT_MAX;

#ifdef _OPENMP
#pragma omp parallel
    {
#ifdef VERBOSE
        int k = omp_get_num_threads();
        int i = omp_get_thread_num();

#pragma omp single
        printf ("OpenMP enabled, running with %d threads\n", k);
#endif
    }
#endif

#pragma omp parallel for
    for (int i=0; i<N; i++) {
        a[i] = 2.0;
        b[i] = 1.0;
        c[i] = 0.8;
        d[i] = 1.01;
    }

    iter = 5;
    times[0] = 0.0;
    times[1] = 0.0;

    while ( times[0] < 0.3 ){
        times[0] = func(a, b, c, d, N, iter);
        if ( times[0] > 0.1 ) break;
        double factor = 0.3 / (times[0] - times[1]);
        iter *= (int) factor;
        times[1] = times[0];
    }

    for ( int k=0; k < NTIMES; k++) {
        times[k] = func(a, b, c, d, N, iter);
    }

    for (int k=1; k<NTIMES; k++) {
        avgtime = avgtime + times[k];
        mintime = MIN(mintime, times[k]);
        maxtime = MAX(maxtime, times[k]);
    }

    double kB = (double) 4.0 * N * bytesPerWord;
    double flops = (double) 2.0 * N * iter * scale;
    printf("%.2f %.2f\n", 1.0E-03 * kB, 1.0E-06 * flops/mintime);

    return EXIT_SUCCESS;
}

double getTimeStamp()
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec * 1.e-9;
}

double striad_seq(
        double * restrict a,
        const double * restrict b,
        const double * restrict c,
        const double * restrict d,
        int N,
        int iter
        )
{
    double S, E;

    S = getTimeStamp();
    for(int j = 0; j < iter; j++) {
#pragma vector aligned
        for (int i=0; i<N; i++) {
            a[i] = b[i] + d[i] * c[i];
        }

        if (a[N-1] > 2000) printf("Ai = %f\n",a[N-1]);
    }
    E = getTimeStamp();

    return E-S;
}

double striad_tp(
        double * restrict a,
        const double * restrict b,
        const double * restrict c,
        const double * restrict d,
        int N,
        int iter
        )
{
    double S, E;

#pragma omp parallel
    {
        double* al;
        posix_memalign((void**) &al, ARRAY_ALIGNMENT, N * sizeof(double));

#pragma omp single
        S = getTimeStamp();
        for(int j = 0; j < iter; j++) {
#pragma vector aligned
            for (int i=0; i<N; i++) {
                al[i] = b[i] + d[i] * c[i];
            }

            if (al[N-1] > 2000) printf("Ai = %f\n",al[N-1]);
        }
#pragma omp single
        E = getTimeStamp();
    }

    return E-S;
}

double striad_ws(
        double * restrict a,
        const double * restrict b,
        const double * restrict c,
        const double * restrict d,
        int N,
        int iter
        )
{
    double S, E;

    S = getTimeStamp();
#pragma omp parallel
    {
        for(int j = 0; j < iter; j++) {
#pragma omp for
#pragma vector aligned
            for (int i=0; i<N; i++) {
                a[i] = b[i] + d[i] * c[i];
            }
            if (a[N-1] > 2000) printf("Ai = %f\n",a[N-1]);
        }
    }
    E = getTimeStamp();

    return E-S;
}

File: StreamTriad/Julia/streamTriad.jl
#!/usr/bin/env julia
# =======================================================================================
#
#      Author:   Thomas Gruber (tg), thomas.gruber@fau.de
#      Copyright (c) 2020 RRZE, University Erlangen-Nuremberg
#
#      Permission is hereby granted, free of charge, to any person obtaining a copy
#      of this software and associated documentation files (the "Software"), to deal
#      in the Software without restriction, including without limitation the rights
#      to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#      copies of the Software, and to permit persons to whom the Software is
#      furnished to do so, subject to the following conditions:
#
#      The above copyright notice and this permission notice shall be included in all
#      copies or substantial portions of the Software.
#
#      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#      IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#      FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#      AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#      LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#      OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#      SOFTWARE.
#
# =======================================================================================

using Printf

const NTIMES = Int64(10)


function striad_seq(A::Array{Float64,1}, B::Array{Float64,1}, C::Array{Float64,1}, D::Array{Float64,1}, vsize::Int64, iters::Int64)
    for i in 1:iters
        @inbounds for j in 1:vsize
            A[j] = B[j] + C[j] * D[j]
        end
    end
    return nothing
end


function striad_ws(A::Array{Float64,1}, B::Array{Float64,1}, C::Array{Float64,1}, D::Array{Float64,1}, vsize::Int64, iters::Int64)
    for i in 1:iters
        Threads.@threads for j in 1:vsize
            @inbounds A[j]= B[j] + C[j] * D[j]
        end
    end
    return nothing
end

function striad_tp(A::Array{Float64,1}, B::Array{Float64,1}, C::Array{Float64,1}, D::Array{Float64,1}, vsize::Int64, iters::Int64)
    tasks = Vector{Task}(undef, Threads.nthreads())
    for t in 1:Threads.nthreads()
        AL = zeros(Float64, vsize)
        tasks[t] = Threads.@spawn striad_seq(AL, B, C, D, vsize, iters)
    end
    for t in 1:Threads.nthreads()
        wait(tasks[t])
    end
end

function usage()
    @printf("Usage: %s <test type>  <N>\n", PROGRAM_FILE)
    println("Test types: 0 - sequential, 1 - throughput, 2 - worksharing")
end

function main()
    typ = 0
    VECTORSIZE:Int64 = 1
    iters::Int64 = 5
    func = striad_seq


    if size(ARGS, 1) != 2
        usage()
        exit(0)
    else
        if ARGS[1] == "0"
            func = striad_seq
            typ = 0
        elseif ARGS[1] == "1"
            func = striad_tp
            scale = Threads.nthreads()
            typ = 1
        elseif ARGS[1] == "2"
            func = striad_ws
            typ = 2
        else
            @printf("Unknown test type: %s\n", ARGS[1])
            exit(1)
        end
        VECTORSIZE = parse(Int64, ARGS[2])
    end

    A = zeros(Float64, VECTORSIZE)
    B = zeros(Float64, VECTORSIZE)
    C = zeros(Float64, VECTORSIZE)
    D = zeros(Float64, VECTORSIZE)

    avgtime = 0
    mintime = 10000000
    maxtime = 0

    for j in 1:VECTORSIZE
        A[j] = 2.0
        B[j] = 1.0
        C[j] = 0.8
        D[j] = 1.01
    end

    times = zeros(Float64, NTIMES)

    while times[1] < 0.3
        times[1] = @elapsed func(A, B, C, D, VECTORSIZE, iters)
        if times[1] > 0.1 break end
        factor = 0.3 / (times[1] - times[2])
        iters = iters * floor(factor)
        times[2] = times[1]
    end

    for k in 1:NTIMES
        times[k] = @elapsed func(A, B, C, D, VECTORSIZE, iters)
    end

    for k in 2:NTIMES
        avgtime = avgtime + times[k]
        mintime = min(mintime, times[k])
        maxtime = max(maxtime, times[k])
    end

    kB = 4.0 * VECTORSIZE * sizeof(Float64)
    flops = 2.0 * VECTORSIZE * iters

    @printf("%.2f %.2f\n", 1.0E-03 * kB, 1.0E-06 * flops/mintime)
end

main()
